src/eve/utils.py

```diff
@@ -555,37 +555,54 @@ def as_dict(self) -> Dict[str, T]:
     asdict = as_dict
 
 
-class FrozenNamespace(Namespace[T]):
-    """An immutable version of :class:`Namespace`.
+class ConstantNamespace(Namespace[T]):
+    """A version of :class:`Namespace` for collecting constant expressions.
 
     Examples:
-        >>> ns = FrozenNamespace(a=10, b="hello")
+        >>> ns = ConstantNamespace(a=10)
+        >>> ns.b = "hello"
         >>> ns.a
         10
         >>> ns.a = 20
         Traceback (most recent call last):
            ...
-        TypeError: Trying to modify immutable 'FrozenNamespace' instance.
+        TypeError: Trying to modify immutable member 'a' of a 'ConstantNamespace'.
 
-        >>> ns = FrozenNamespace(a=10, b="hello")
+        >>> ns = ConstantNamespace(a=10, b="hello")
         >>> list(ns.items())
         [('a', 10), ('b', 'hello')]
+    """
+
+    __slots__ = "__cached_hash_value__"  # This slot is used to avoid polluting the namespace
+
+    def __setattr__(self, __name: str, __value: Any) -> None:
+        if hasattr(self, __name):
+            raise TypeError(
+                f"Trying to modify immutable member '{__name}' of a '{self.__class__.__name__}'."
+            )
+        self.__dict__[__name] = __value
+
+    def __delattr__(self, __name: str) -> None:
+        raise TypeError(f"Trying to modify immutable '{self.__class__.__name__}' instance.")
+
 
-        >>> ns = FrozenNamespace(a=10, b="hello")
+class FrozenNamespace(ConstantNamespace[T]):
+    """
+    A `ConstantNamespace` that cannot be modified by adding new constant attributes.
+
+    Unlike ConstantNamespaces, FrozenNamespaces are hashable.
+
+    Examples:
+        >>> ns = ConstantNamespace(a=10, b="hello")
```

Comment:
This doctest is using a ConstantNamespace and thus should be currently failing (otherwise there is a bug)..

```suggestion
        >>> ns = FrozenNamespace(a=10, b="hello")
```

Reply:
Fixed, thanks!